apm-agent-core/src/test/java/co/elastic/apm/agent/MockTracer.java

```diff
@@ -81,6 +84,47 @@ public static ElasticApmTracer createRealTracer(Reporter reporter, Configuration
             .build();
     }
 
+    /**
+     * If an instrumentation has already been initialized by some other test, returns the static
```

Comment:
Why do we allow tests to leak the tracer? It may be configured to only use certain instrumentations.

We could keep a stack trace of where the tracer initialized. That way we can find out which test has leaked the tracer and fix it instead of re-using a tracer that is in an unknown state.

Reply:
The basic problem is the way we spread around tracer's references, so it is held as a field in many places (all Instrumentations, numerous helpers etc.), which means it leaks by design whenever multiple instrumentation tests are executed on the same JVM. Currently this is only a problem in tests, but it may become a real issue when we implement detach/re-attach.

I think the proper solution is to refactor the tracer to be a singleton, then we can easily control all accesses centrally.